### ISSUE TITLE
Add context argument to grammar recognizers

### ISSUE THREAD

**Issue by jwcraftsman** [igordejanovic/parglare#50]
It is currently possible to extend the LR parser by keeping track of some extra state in global variables that are examined by custom recognizers and modified by custom actions.  This has proven useful in parsing indentation, but probably has many other applications as well.  It would be nice to have the parser keep track of this extra state and pass it along to both the recognizers and the actions.

Phase I: As mentioned in #5, there is a Context object that can be passed as an argument to the the parse() method.  This is currently passed along to actions, but not to recognizers.  This context object could be used to store any extra state variables that are required.  If this context object were passed to recognizers as well, then the external global state could be eliminated.

Phase II: If possible, it would be very useful to make this work for GLR as well.  It seems like the context object, or a particular attribute of it, could be duplicated when GLR parser forks, keeping a separate copy for each fork.  Perhaps copy.deepcopy() or some sort of custom clone() method on the extra state object could be used to perform this duplication.

Question: If this proposal seems acceptable, should the extra context argument be added to the beginning or end of the recognizer argument list?  The actions include it at the beginning, but it might cause less trouble for existing code to add it at the end.

I have started experimenting with this, and have implemented some of the more straightforward changes for LR parsing here:

https://github.com/codecraftingtools/parglare/tree/recognizer-context

Please take a look at it and let me know what you think.  I think this covers most of what is required for phase I.

Another question: In implementing indentation parsing, I needed to allow recognizers to recognize the empty string.  This required modifying this line in _token_recognition()

```
             last_prior = symbol.prior
             tok = symbol.recognizer(input_str, position, context)
-            if tok:
+            if tok is not None:
                 tokens.append(Token(symbol, tok))
```

to differentiate between an empty string ("") and no match (None).  This change is included as a separate commit in the branch mentioned above.  It seems like this small change makes things work, but it may break some other things I am not aware of, so I wanted to point it out.

Any comments would be appreciated.  Also, please let me know if you would like me to do anything differently to make the collaboration workflow easier.  Thanks again for all the work that has gone into this project.

**Comment by igordejanovic** [comment]
Hi Jeff. Thanks for you work and interest. 
Overall, the changes on your branch looks good. Yes, I think making context as the last parameter is easier to deal with. I'm thinking even making it optional for both actions and recognizers as most of the time it's not needed. Something like:

```python
try:
   # call action or recognizer without context
except TypeError:
   # call it with context
``` 
This might introduce problems if `TypeError` is raised by the actual user code though.

I think your change in `_token_recognition` is actually a fix as recognition should be considered unsuccessful *only* if `None` is returned.

Please keep in mind that parglare can parse not only strings but any Python sequence (e.g. list of objects), Please see [here](https://github.com/igordejanovic/parglare/blob/master/tests/func/test_recognizers.py). So token recognizers may return recognized sub-lists. Also, in that case, layout is a list too.

Cloning context in GLR might be a good idea. I was going the route of reusing the same object due to performance reasons but not sure that there is much of a gain achieved given the number of assignments needed to reconstruct the state of the context at each iteration.

**Comment by igordejanovic** [comment]
Regarding my comment on making context param optional. A safer approach might be to use introspection to see how many parameters are there. Since that would be a significant overhead it should be done only once during parser construction and all actions/recognizers should be marked (e.g. by attaching attribute `_pg_context = True`). Later it would be trivial to call actions/recognizers with the right parameters.

**Comment by jwcraftsman** [comment]
Thanks for the input.  I initially implemented the try/except approach for handling the optional context parameter (similar to what you suggested) to avoid changing the existing recognizers.  I ended up changing it to what you see now because (as usually happens when I do that sort of thing) I made some mistake and the try/except was masking the true exception.  I haven't used the introspection approach before, but that sounds like a robust way to handle this.  I will look into that.

I will also take a look at how to implementing cloning of the context for GLR.  It sounds like you think that cloning the whole object would be worth a look.  If you didn't want to do that, I thought that we might could just specify a certain attribute that would be cloned.  I think cloning the whole context as you suggested would be cleaner, though, if you think the performance would be okay.

**Comment by jwcraftsman** [comment]
I took a swag at implementing an optional leading context argument for recognizers in my branch.  This was done using introspection and attaching a _pg_context attribute to the recognizers, as was suggested.  Please let me know if I left something out, put the check in the wrong place, or if we need to take a different approach.

Should I break out the _token_recognition "is not None" fix into a separate issue, or merge it in with the rest of these changes if/when the time comes?

What else needs fixing in my branch aside from the possibly implementing GLR context cloning?

**Comment by jwcraftsman** [comment]
For GLR, would the context object be cloned and then stored in a GSSNode, replacing the existing start_position, end_position, and layout_content GSSNode members?

**Comment by jwcraftsman** [comment]
The reason I used getargspec() instead of signature() was to support Python 2.  I didn't _test_ my code under Python 2, though, and it turns out that it works a little differently.  I committed a quick fix for this that works for both Python 2 and 3, although I can update it to use the approach you mentioned above.

**Comment by igordejanovic** [comment]
According to [this](https://stackoverflow.com/questions/847936/how-can-i-find-the-number-of-arguments-of-a-python-function) `getargspec` is deprecated for Python 3 so it would be better to with `signature` in that case.

**Comment by jwcraftsman** [comment]
I have updated my branch according to your suggestions.  Terminals now have a recognizer property that checks the signature in the setter function using inspect.signature for Python 3 and getargspec for Python 2.  I also created a new pull request for the "recognizer return value checking" fix.

Since performance seems to be a big issue, maybe just having a specific property on the GSSNodes that gets cloned would be better.  A added an "extra" property that gets cloned using deepcopy.  This should minimize the performance hit if extra state is not required.  I don't follow all of the GLR logic, so I probably missed some saving/restoring of the extra property.  Someone else may need to finish that part of the implementation, but it seems like it's getting close.  The committed changes seemed to work for this example:

```
#!/usr/bin/env python3

import re
from parglare import Parser, GLRParser, Grammar
from parglare.parser import Context

grammar_string = r"""
terms: term* EOF;
term: term1 | term2;

terminals
term1: ;
term2: ;
"""

N = 4

class Extra(object):
    def __init__(self):
        self.term_count = 0

term_re = re.compile(r"[a-z]+")
def recognize_term1(context, input, pos):
    match = term_re.match(input, pos)
    if match:
        return input[pos:match.end()]

term_re = re.compile(r"[a-z]+")
def recognize_term2(context, input, pos):
    if context.extra.term_count >= N:
        return None
    match = term_re.match(input, pos)
    if match:
        return input[pos:match.end()]

recognizers = {
    "term1": recognize_term1,
    "term2": recognize_term2,
}

grammar = Grammar.from_string(grammar_string, recognizers=recognizers)

def term1_action(context, node):
    context.extra.term_count += 1
    return 1

def term2_action(context, node):
    context.extra.term_count += 1
    return 2

actions = {
    "term1": term1_action,
    "term2": term2_action,
}

input_string = r"""
a b c d e f
"""

parser = GLRParser(grammar, actions=actions)

ctx = Context()
ctx.extra = Extra()
r = parser.parse(input_string, context=ctx)
for rr in r:
    print(rr)
```
The current branch seems to work for both Python 3 and Python 2.  Please review and let me know what you think about where things stand.

**Comment by jwcraftsman** [comment]
> The same approach can be used for actions so we retain backward compatibility while simplifying API.

What needs to be done regarding actions?

**Comment by jwcraftsman** [comment]
Another plus of having a special context attribute reserved for user state is that we don't have to worry about adding more parglare contexts attributes in the future and having name collisions with user state.

**Comment by igordejanovic** [comment]
> What needs to be done regarding actions?

Context is rarely needed for actions either so that parameter could be made optional also.

> Another plus of having a special context attribute reserved for user state is that we don't have to worry about adding more parglare contexts attributes in the future and having name collisions with user state.

Agreed.

**Comment by jwcraftsman** [comment]
Should we make the context parameter for actions optional in this enhancement, or should that be addressed in a separate issue?  When this change is made, I assume the context parameter should removed for built-in actions that don't use it.

**Comment by igordejanovic** [comment]
> Should we make the context parameter for actions optional in this enhancement, or should that be addressed in a separate issue? 

I think a separate issue/PR would be easier to review.

>  When this change is made, I assume the context parameter should removed for built-in actions that don't use it.

Yes. Also from all tests and examples.

**Comment by jwcraftsman** [comment]
I see that 45b6f94c6f17b1f88af21aa27ad9cdc280c36a56 breaks the example I gave in the comment above (https://github.com/igordejanovic/parglare/issues/50#issuecomment-407998662)

Now I get this erroneous result:
```
[[1, 1, 1, 1, 1, 1], None]
[[2, 1, 1, 1, 1, 1], None]
[[1, 2, 1, 1, 1, 1], None]
[[2, 2, 1, 1, 1, 1], None]
```
when the previous commit yields:
```
[[1, 1, 1, 1, 1, 1], None]
[[2, 1, 1, 1, 1, 1], None]
[[1, 2, 1, 1, 1, 1], None]
[[2, 2, 1, 1, 1, 1], None]
[[1, 1, 2, 1, 1, 1], None]
[[2, 1, 2, 1, 1, 1], None]
[[1, 2, 2, 1, 1, 1], None]
[[2, 2, 2, 1, 1, 1], None]
[[1, 1, 1, 2, 1, 1], None]
[[2, 1, 1, 2, 1, 1], None]
[[1, 2, 1, 2, 1, 1], None]
[[2, 2, 1, 2, 1, 1], None]
[[1, 1, 2, 2, 1, 1], None]
[[2, 1, 2, 2, 1, 1], None]
[[1, 2, 2, 2, 1, 1], None]
[[2, 2, 2, 2, 1, 1], None]
```
What was the reason for this change?  Was deepcopy was too much of a performance hit?  Is there a solution, or is GLR not supported in this case?

**Comment by igordejanovic** [comment]
Thanks for pointing that out. 

I went to shallow copy as I realized that deepcopy would be a half-baked solution as the context is kept on GSSNode which represent a merged state of multiple parser threads. I feel that it could be a dangerous to assume that the extra context is a separate value per GLR parser thread as, during heads merge, parser has to decide which extra value to keep on the merged head. The proper solution would be to merge the heads only if extra values are equal but that could *possibly* lead to the exponential increase of the number of parser active heads as the otherwise legit merges would be prevented by this equality check.

Until I get a clearer view on the subject and find a proper way to keep extra value per GLR parser thread I decided it would be a safer approach to introduce extra as a shallow copied value. Think of it as a global value for the whole parser. Similar you could get by attaching a state attributes to the `context.parser`. Not ideal but should serve the purpose until a better solution is found.

**Comment by igordejanovic** [comment]
Fixed in #55. Available in 0.7.0

**Comment by jwcraftsman** [comment]
That makes sense.  Thanks for creating another issue as a reminder to look for a solution.
